### PR TITLE
Add webhook to prevent the prometheusrule modification in managed nam…

### PR DESCRIPTION
**Pull request by reetika-vyas**
I will close OSD-10127, It will block users with cluster-admin privileges  from creating prometheusrule in managed-namespace.

### CODE REVIEW

File: pkg/webhooks/prometheusrule/prometheusrule.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	admissionctl "sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	WebhookName string = "prometheusrule-validation"
+	docString   string = `Managed OpenShift Customers may not create PrometheusRule in namespaces managed by Red Hat .`
```
Comment (lisa) [inline pkg/webhooks/prometheusrule/prometheusrule.go:20]:
```suggestion
	docString   string = `Managed OpenShift Customers may not create PrometheusRule in namespaces managed by Red Hat.`
```

Unneeded white space.

**Comment by feichashao** [comment]
/hold
 before merge, we want to see how many clusters are having such issue, ref https://issues.redhat.com/browse/OSD-10127?focusedCommentId=20961163&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-20961163

**Comment by feichashao** [comment]
@lisa  I see @reetika-vyas has addressed the suggestion, could you help to take another look?

**Comment by lisa** [comment]
> @lisa I see @reetika-vyas has addressed the suggestion, could you help to take another look?

Yep, that particular chnage looks fine. 👍

File: pkg/webhooks/prometheusrule/prometheusrule_test.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+			shouldBeAllowed: true,
+		},
+		{
+			testID:          "regular-user-can-create-prometheusrule-in-managed-namespaces",
+			targetNamespace: "openshift-user-workload-monitoring",
+			targetResource:  "prometheusrule",
+			username:        "user3",
+			userGroups:      []string{"system:authenticated", "system:authenticated:oauth"},
+			operation:       admissionv1.Update,
+			shouldBeAllowed: true,
```
Comment (feichashao) [inline pkg/webhooks/prometheusrule/prometheusrule_test.go:275]: what are the difference between these 3 test cases?  the testID should mean the user can create prometheusrule in openshift-user-workload-monitoring but not other managed namespaces?

File: pkg/webhooks/prometheusrule/prometheusrule.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+func (s *prometheusruleWebhook) authorized(request admissionctl.Request) admissionctl.Response {
+	var ret admissionctl.Response
+
+	pr, err := s.renderPrometheusRule(request)
+	if err != nil {
+		log.Error(err, "Couldn't render a PrometheusRule from the incoming request")
+		return admissionctl.Errored(http.StatusBadRequest, err)
+	}
+
+	if hookconfig.IsPrivilegedNamespace(pr.GetNamespace()) && pr.GetNamespace() != "openshift-user-workload-monitoring" {
```
Comment (wanghaoran1988) [inline pkg/webhooks/prometheusrule/prometheusrule.go:79]: the Privileged namespaces is a long list including all addon namespaces, will this block addon components to create the rule?

Comment (reetika-vyas) [reply pkg/webhooks/prometheusrule/prometheusrule.go:82]: I have added the changes requested. Can I get LGTM on this? @wanghaoran1988

File: pkg/webhooks/prometheusrule/prometheusrule_test.go (hunk 2 above)
Comment (reetika-vyas) [reply pkg/webhooks/prometheusrule/prometheusrule_test.go]: yes

File: pkg/webhooks/prometheusrule/prometheusrule.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	admissionctl "sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	WebhookName               string = "prometheusrule-validation"
+	docString                 string = `Managed OpenShift Customers may not create PrometheusRule in namespaces managed by Red Hat.`
+	privilegedServiceAccounts string = `^system:serviceaccounts:(kube.*|openshift.*|default|redhat.*|osde2e-[a-z0-9]{5})`
```
Comment (wanghaoran1988) [inline pkg/webhooks/prometheusrule/prometheusrule.go:22]: s/privilegedServiceAccounts/privilegedServiceAccountGroups/

File: pkg/webhooks/prometheusrule/prometheusrule_test.go [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+			testID:          "regular-user-can-create-prometheusrule-in-openshift-user-workload-monitoring",
+			targetNamespace: "openshift-user-workload-monitoring",
+			targetResource:  "prometheusrule",
+			username:        "user3",
+			userGroups:      []string{"cluster-admins", "system:authenticated", "system:authenticated:oauth"},
+			operation:       admissionv1.Update,
+			shouldBeAllowed: true,
+		},
+		{
+			testID:          "regular-user-can-create-prometheusrule-in-redhat-rhoam-observability",
```
Comment (wanghaoran1988) [inline pkg/webhooks/prometheusrule/prometheusrule_test.go:278]: serviceaccount-in-managed-namespaces-can-create....

File: pkg/webhooks/prometheusrule/prometheusrule_test.go [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+			username:        "system:serviceaccounts:redhat-ns",
+			userGroups:      []string{"system:serviceaccounts:redhat-ns", "system:authenticated", "system:authenticated:oauth"},
+			operation:       admissionv1.Create,
+			shouldBeAllowed: true,
+		},
+		{
+			testID:          "regular-user-can-update-prometheusrule-in-redhat-rhoam-observability",
+			targetNamespace: "redhat-rhoam-observability",
+			targetResource:  "prometheusrule",
+			username:        "system:serviceaccounts:redhat-ns",
```
Comment (wanghaoran1988) [inline pkg/webhooks/prometheusrule/prometheusrule_test.go:290]: username should be something like "system:serviceaccounts:redhat-ns:test-operator"

**Review by NautiluX — COMMENTED** [review]
/hold cancel

**Pull request merged**